modelgauge/pipeline.py

```diff
@@ -0,0 +1,310 @@
+
+ In this code, a Pipeline is a collection of PipelineSegments, each of which runs in its own threads. A Source
+ is the head of the pipe; a Pipe is any number of middle processing stages; a Sink is the last stage. A Pipe
+ can have multiple threads, which is best used when when python is waiting for something external to happen,
+ like a remote API call.
+
+ To create your own pipeline, you'll need to subclass Source, Pipe (possibly more than once), and Sink, implementing
+ the abstract method on each. Then you can put them together like this:
+
+    p = Pipeline(
+        MySource(),
+        MyPipe(),
+        MySink()
+    )
+    p.run()
+
+That will run every item produced by the source, passing the output of each stage along to the next.
+
+What gets passed along? Whatever your stages produce. There's no type requirement. So for example:
+
+    class MySource(Source):
+
+        def new_item_iterable(self):
+            return [1, 2, 3]
+
+
+    class MyPipe(Pipe):
+
+        def handle_item(self, item):
+            return item * 2
+
+
+    class MySink(Sink):
+
+        def handle_item(self, item):
+            print(item)
+
+Will print 2, 4, and 6 when run.
+
+That example just has one thread per PipeSegment. If you'd like more workers in a given Pipe, you might do something
+like:
+
+    class MyPipe(Pipe):
+
+        def __init__(self, ):
+            super().__init__(thread_count=8)
+
+Note that this generally only helps when the code is waiting on the network or other I/O. Python's Global Intepreter
+Lock (GIL) means that generally only one bit of python is running at once.
+
+ """
+import datetime
+import queue
+import sys
+import threading
+from abc import abstractmethod, ABC
+from queue import Queue
+from threading import Event, Thread
+from typing import Any, Callable, Iterable, Optional
+
+
+class PipelineSegment(ABC):
+    """A segment of a Pipeline used for parallel processing."""
+
+    default_timeout = 0.1
+
+    def __init__(self):
+        super().__init__()
+        self._work_done = Event()
+        self._upstream: Optional[PipelineSegment] = None
+        self._queue: Queue = Queue()
+        self.completed = 0
+        self._debug_enabled = False
+
+    def start(self):
+        pass
+
+    def downstream_put(self, item: Any):
+        """Enqueue an item for the next segment in the pipeline to take."""
+        self._queue.put(item)
+
+    def upstream_get(self, timeout=None):
+        """Dequeue an item from the previous segment in the pipeline"""
+        if timeout is None:
+            timeout = PipelineSegment.default_timeout
+        if self._upstream is None:
+            raise (ValueError(f"{self} doesn't have an upstream connection"))
+        self._debug("getting from upstream")
+
+        try:
+            item = self._upstream._queue.get(timeout=timeout)
+        except queue.Empty:
+            self._debug("get was empty")
+            raise
```

Comment:
I'm curious why would this be an error

Reply:
This is one of those cases where people are using exceptions for things that are not errors. If you call a Queue with a timeout and nothing arrives, it will raise `queue.Empty` as part of its normal operations. Personally, I would have it just return `None`, but I think they did it this way because they wanted people to be able to put `None` into the queue, or at least not to have to worry if they were putting `None` in. 

Here I want to have a good debug message in this case, so I did the catch-and-raise thing. One layer up there's something that catches `queue.Empty` and handles it properly.

Reply:
Okay I think I understand now, thanks for explaining!